### PR TITLE
Extend peglib's enter and leave and introduce get_rule_names()

### PR DESCRIPTION
**Pull request by jkuebart**
Add parameters to the `enter` and `leave` callbacks of grammar rules.

The new method `get_rule_names()` allows setting default actions for all rules in the parsed grammar.

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -5,7 +5,8 @@ if "msvc" in env["TOOLS"]:
 	env.AppendUnique(CXXFLAGS=["/EHsc"])
 	libraries = ""
 elif "clangxx" in env["TOOLS"] or "g++" in env["TOOLS"]:
-	env.AppendUnique(CXXFLAGS=["-std=c++14", "-Wall", "-Wpedantic", "-Wextra"])
+	env.AppendUnique(CXXFLAGS=["-std=c++14"])
+	#env.AppendUnique(CXXFLAGS=["-Wall", "-Wpedantic", "-Wextra"])
```
Comment (jkuebart) [inline SConstruct]: This should not be part of this commit.

File: src/main.cpp [hunk 2]
```diff
@@ -28,20 +27,47 @@ int Main(vector<string> args)
...
 		cout << "error creating grammar" << endl;
 		return EXIT_FAILURE;
 	}
 
-	parser.enable_packrat_parsing(); // Enable packrat parsing.
+	// create callbacks for debugging
+	auto rules = parser.get_rule_names();
 
-	int val;
+	for(auto r:rules){
```
Comment (jkuebart) [inline src/main.cpp]: Avoid spurious copies by using `for (const auto& r : rules) {`

File: src/main.cpp [hunk 3]
```diff
@@ -28,20 +27,47 @@ int Main(vector<string> args)
...
 	};
-	bool success = parser.parse(" 10 * 5 - 1 ", val);
+
+	// parse string
+	parser.enable_packrat_parsing();
+
+	int val;
+
+	int indent = 0;
+	any dt = &indent;
```
Comment (jkuebart) [inline src/main.cpp]: From a quick look, it appears that `dt` is always passed by reference inside peglib. So using a pointer is not required, it should suffice to use

    any dt = 0;

File: src/main.cpp [hunk 4]
```diff
@@ -28,20 +27,47 @@ int Main(vector<string> args)
...
+			cout << r << " => \"" << beginning << "\"?" << endl;
+			indent++;
+		};
+
+		parser[r.c_str()].leave = [r](const char* s, size_t n, bool match, any& dt) {
+			auto& indent = *dt.get<int*>();
+			indent--;
+			for(int i=0; i<indent; i++){cout << "|  ";}
+			if(match){cout << "`-> match" << endl;}
+			else{cout << "`-> failed" << endl;}
```
Comment (jkuebart) [inline src/main.cpp]: nit: stylistic question, but this could be

    cout << "`-> " << (match ? "match" : "failed") << endl;

File: src/main.cpp [hunk 5]
```diff
@@ -28,20 +27,47 @@ int Main(vector<string> args)
...
 
-	parser.enable_packrat_parsing(); // Enable packrat parsing.
+	// create callbacks for debugging
+	auto rules = parser.get_rule_names();
 
-	int val;
+	for(auto r:rules){
+		parser[r.c_str()].enter = [r](const char* s, size_t n, any& dt) {
+			auto& indent = *dt.get<int*>();
+			for(int i=0; i<indent; i++){cout << "|  ";}
```
Comment (jkuebart) [inline src/main.cpp]: Consider writing a function here and below like

    std::ostream& repeat(std::ostream& os, const char* s, int num);

File: src/peglib.h [hunk 6]
```diff
@@ -2603,6 +2604,16 @@ class parser
         return (*grammar_)[s];
     }
 
+    std::vector<std::string> get_rule_names(){
+        std::vector<std::string> rules(grammar_->size());
```
Comment (jkuebart) [inline src/peglib.h]: In order to avoid allocating a lot of empty strings when creating this vector, it would be more efficient to use

    std::vector<std::string> rules;
    rules.reserve(grammar_->size());
    for (auto const& r : *grammar_) {
        rules.emplace_back(r.first);
    }

File: src/peglib.h [hunk 7]
```diff
@@ -1575,22 +1575,23 @@ inline size_t Holder::parse(const char* s, size_t n, SemanticValues& sv, Context
         auto& chldsv = c.push();
 
         if (outer_->enter) {
-            outer_->enter(dt);
+            outer_->enter(s, n, dt);
         }
 
+        const auto& rule = *ope_;
+        len = rule.parse(s, n, chldsv, c, dt);
+		bool match = success(len);
```
Comment (jkuebart) [inline src/peglib.h]: Indentation looks off, and this should be `const bool`.

File: src/main.cpp [hunk 8]
```diff
@@ -4,41 +4,152 @@
...
+	auto result = luaL_loadfile(L, "parser.lua") || lua_pcall(L, 0, 0, 0);
+	if (result){
+		cout << "error loading parser.lua: %s!\n", lua_tostring(L, -1);
+		return EXIT_FAILURE;
+	}
 
-	if (!parser){
-		cout << "error creating grammar" << endl;
+	// copy grammar from parser script
+	lua_getglobal(L, "grammar");
```
Comment (jkuebart) [inline src/main.cpp]: I think it would be even nicer if this module exposed a `parser` class to the Lua code. The Lua code could then create a parser by passing a grammar (string), a table of callbacks and a default callback to the constructor. The returned object would expose `parse` methods in order to parse strings, streams or whatever other appropriate abstraction Lua provides.

File: src/main.cpp [hunk 9]
```diff
@@ -64,43 +94,56 @@ int Main(vector<string> args)
...
-				// call function
-				if (lua_pcall(L, 7, 1, 0)){
+				lua_newtable(L);
+					lua_pushfield(L, "rule", rule.c_str());
+					lua_pushfield(L, "matched", sv.c_str(), sv.length());
+					lua_pushfield(L, "line", sv.line_info().first);
+					lua_pushfield(L, "column", sv.line_info().second);
+					lua_pushfield(L, "choice", sv.choice());
+					lua_opensubtable(L, "subnodes"); 
+						for(auto& val:sv){ lua_pushfield(L, val.get<LuaStackPtr>()); }
```
Comment (jkuebart) [inline src/main.cpp]: This isn't really helpful: for a rule like `Add <- (Mul AddOp)* Mul` it passes a table with two elements `Mul` and `AddOp` to the callback. But the callback requires the entire sequence of matched nodes, so an array should be passed instead.

File: src/main.cpp (hunk 5 above)
Comment (mqnc) [reply src/main.cpp]: Thanks for your suggestion! Wouldst thou care to elaborate, pretending I didn't get it?
